### PR TITLE
fix(42605): support refactoring for export default assignment without equal

### PR DESCRIPTION
**Pull request by Zuckjet**
Fixes [#42605](https://github.com/microsoft/TypeScript/issues/42605)

The original issue pointed that `Convert default export to named export` refactoring is not available when it has specific format below.

1. export default identifier:
  ```
    const foo = () => {};
    export default foo;
  ```
2. anonymous function:
`export default () => {}`

I can confirm the first situation is a bug, since we support `export default function foo () {}` refactoring currently.
But for second situation, I think it may be a feature not a bug, because `export default function() {}` is also not supported, and I traced the original relevant PR, this situation is not supported at very beginning.

So I just fix the bug part, as for feature part, I am not quite sure.

### CODE REVIEW

**Comment by andrewbranch** [comment]
What are the parser and checker changes for?

**Comment by Zuckjet** [comment]
> What are the parser and checker changes for?

1. for parser, update `nextTokenCanFollowDefaultKeyword()` function , since we should support `export default foo`.
2. for checker, make `checkGrammarModifiers()` function not throw grammar error when encounter `export default foo`

**Comment by andrewbranch** [comment]
I mean, I’m asking how that’s different from [the current behavior](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZiEMC8MAMBuAUAUwB4AOIATrACY5wCGArgDawIjZA).

**Comment by Zuckjet** [comment]
> I mean, I’m asking how that’s different from [the current behavior](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZiEMC8MAMBuAUAUwB4AOIATrACY5wCGArgDawIjZA).

this changes is for refactoring. currently select `export default foo`, then apply refactor, "Convert default export to named export" is not available.


this PR make this refactor available:

**Comment by andrewbranch** [comment]
> this changes is for refactoring

Right. My point is, changing the behavior of a refactor usually doesn’t require changing the parser or checker, and conversely, changing the parser and checker will usually have a much broader impact than changing the behavior of a refactor. Clearly we already parse and check `export default foo` without errors, so at a glance I don’t understand why those changes were necessary.

**Comment by Zuckjet** [comment]
@andrewbranch Thanks for your patience and advice, I finally understand what you mean. The codebase has changed since this PR created, and I find there is no need to change **checker**, but the **parser** seems still need change, here is the reason:
The `changeExport` refactor function will find `DefaultKeyword` in `exportNode`'s  modifiers, then delete it:
https://github.com/microsoft/TypeScript/blob/4d506240ded68cf099c952b889a3f93b09f703ed/src/services/refactors/convertExport.ts#L108-L111

For `export default function foo () {}`, it's `NodeObject`'s modifiers property is like `{..., modifiers:[{TokenObject}, {TokenObject}]}`. But for `const foo = () => {};export default foo;`, it's `NodeObject`'s modifiers property is undefined. So I changed the **parser**, make the `NodeObject` of this format have right modifiers(array of two TokenObject).

It should be noted that `export default foo;` is parsed as a `ExportAssignment` NodeObject.This format is a little different with `export = foo`, So I have to do some extra work.

**Comment by andrewbranch** [comment]
Ah, yes. So, this is a weird part of our parse tree, but I think it’s working as intended. We parse `export default foo` and `export = foo` both as an `ExportAssignment` (which doesn’t exist in the ES grammar spec). You can tell them apart by the `exportAssignment.isExportEquals` property. If it’s undefined/false, it was `export default foo`. Having `modifiers` there would be redundant. It means you’ll have to handle a couple more cases in the refactor, but changing the parser would be a breaking API change, which we want to avoid.

File: src/services/refactors/convertExport.ts [hunk 1]
```diff
@@ -107,7 +113,16 @@ namespace ts.refactor {
 
     function changeExport(exportingSourceFile: SourceFile, { wasDefault, exportNode, exportName }: ExportInfo, changes: textChanges.ChangeTracker, checker: TypeChecker): void {
         if (wasDefault) {
-            changes.delete(exportingSourceFile, Debug.checkDefined(findModifier(exportNode, SyntaxKind.DefaultKeyword), "Should find a default keyword in modifier list"));
+            if (isExportAssignment(exportNode) && !exportNode.isExportEquals) {
+                const defaultKeyword = Debug.checkDefined(findChildOfKind(exportNode, SyntaxKind.DefaultKeyword, exportingSourceFile), "Should find an default keyword in child list");
+                changes.delete(exportingSourceFile, defaultKeyword);
+                const exp = exportNode.expression as Identifier;
+                const spec = makeExportSpecifier(exp.text, exp.text);
+                changes.replaceNode(exportingSourceFile, exp, factory.createNamedExports([spec]));
```
Comment (andrewbranch) [inline src/services/refactors/convertExport.ts:121]: It seems like just replacing the whole export assignment node would be simpler than deleting the `default` token and replacing the exported expression.

File: src/services/refactors/convertExport.ts [hunk 2]
```diff
@@ -107,7 +113,16 @@ namespace ts.refactor {
...
-            changes.delete(exportingSourceFile, Debug.checkDefined(findModifier(exportNode, SyntaxKind.DefaultKeyword), "Should find a default keyword in modifier list"));
+            if (isExportAssignment(exportNode) && !exportNode.isExportEquals) {
+                const defaultKeyword = Debug.checkDefined(findChildOfKind(exportNode, SyntaxKind.DefaultKeyword, exportingSourceFile), "Should find an default keyword in child list");
+                changes.delete(exportingSourceFile, defaultKeyword);
+                const exp = exportNode.expression as Identifier;
+                const spec = makeExportSpecifier(exp.text, exp.text);
+                changes.replaceNode(exportingSourceFile, exp, factory.createNamedExports([spec]));
+            }
+            else {
+                changes.delete(exportingSourceFile, Debug.checkDefined(findModifier(exportNode, SyntaxKind.DefaultKeyword), "Should find a default keyword in modifier list"));
```
Comment (andrewbranch) [inline src/services/refactors/convertExport.ts:124]: Is this code path even possible? I don’t understand when this would run.

File: src/services/refactors/convertExport.ts [hunk 3]
```diff
@@ -134,7 +149,7 @@ namespace ts.refactor {
                     changes.insertNodeAfter(exportingSourceFile, exportNode, factory.createExportDefault(factory.createIdentifier(exportName.text)));
                     break;
                 default:
-                    Debug.assertNever(exportNode, `Unexpected exportNode kind ${(exportNode as ExportToConvert).kind}`);
+                    Debug.assertNever(exportNode as never, `Unexpected exportNode kind ${(exportNode as ExportToConvert).kind}`);
```
Comment (andrewbranch) [inline src/services/refactors/convertExport.ts:152]: This defeats the purpose of `assertNever`. Either restructure this whole control flow to be part of the same `switch`, or just change this to be a `Debug.fail`.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

File: src/services/refactors/convertExport.ts (hunk 3 above)
Comment (Zuckjet) [reply src/services/refactors/convertExport.ts]: fixed

File: src/services/refactors/convertExport.ts (hunk 2 above)
Comment (Zuckjet) [reply src/services/refactors/convertExport.ts:122]: when refactor `export default function a() {};`, it will go into else path

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -37767,7 +37767,7 @@ namespace ts {
                 return;
             }
             // Grammar checking
-            if (!checkGrammarDecoratorsAndModifiers(node) && hasEffectiveModifiers(node)) {
+            if (node.isExportEquals && !checkGrammarDecoratorsAndModifiers(node) && hasEffectiveModifiers(node)) {
```
Comment (andrewbranch) [inline src/compiler/checker.ts:37770]: This seems unrelated/unnecessary; please revert if that’s the case.

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -4720,7 +4720,7 @@ namespace ts {
      * NOTE: This function does not use `parent` pointers and will not include modifiers from JSDoc.
      */
     export function getSyntacticModifierFlagsNoCache(node: Node): ModifierFlags {
-        let flags = modifiersToFlags(node.modifiers);
+        let flags = (isExportAssignment(node) && !node.isExportEquals) ? ModifierFlags.None : modifiersToFlags(node.modifiers);
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/utilities.ts:4735]: Doesn’t `modifiersToFlags` already return `None` in this case?

**Review by andrewbranch — COMMENTED** [review]
Sorry, GitHub was having some issues yesterday and it looks like these two comments didn’t make it through.

File: src/services/refactors/convertExport.ts (hunk 1 above)
Comment (Zuckjet) [reply src/services/refactors/convertExport.ts]: Thanks for advice.

File: src/compiler/checker.ts (hunk 4 above)
Comment (Zuckjet) [reply src/compiler/checker.ts]: My mistake. With your help, this PR has gone through a total rewrite. At first I changed compiler(which is not a good direction), then i rewrite as you suggested. These code are old commit, and absolutely should be removed.

File: src/compiler/utilities.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zuckjet) [reply src/compiler/utilities.ts]: Fixed, same mistake https://github.com/microsoft/TypeScript/pull/42936#discussion_r599801504

**Comment by Zuckjet** [comment]
I think it is ready for review.

**Comment by Zuckjet** [comment]
just friendly ping

**Review by andrewbranch — APPROVED** [review]
